### DISCUSSION TITLE
Does turbo support negative globs for turbo.json task inputs?

### DISCUSSION THREAD

**Discussion by zaripych** [vercel/turborepo#1372] [Ideas]
E.g. in CRA apps tests are in `src/**` directory and should not affect `build`.
```json
    "build": {
      "dependsOn": [],
      "inputs": [
        "src/**/*.ts",
        "src/**/*.tsx"
      ],
      "outputs": ["dist"]
    }
```
I've tried to use negative patterns: `!src/**/__tests__` and `!src/**/*.test.ts` ... but they don't seem to be effective and changes in tests cause `build` task hash to change. I've tried putting them in different order with no luck. There is a brief mention of negative globs for `--ignore` parameter of the CLI ... but I don't think it works here.

**Comment by zaripych** [comment]
Looks like negative globs are not supported for `inputs` as per https://github.com/vercel/turborepo/blob/cfb65cb818fcc3b4773584465c60b83b9e0b9de1/cli/internal/taskhash/taskhash.go#L124 Would it be a terrible idea to add support? To me it seems like we could be something simple in `manuallyHashPackage` that would use `inputs` where globs start with `!` as ignore patterns.

**Reply by zaripych** [reply]
I looked at `git ls-files` pathway of hash computation as well (it would require git style negation using `:!:`) and have a small POC working. Let me know if this is worth a PR.

**Reply by weyert** [reply]
Off topic, but do you have link about :!:?

**Reply by zaripych** [reply]
Here we go https://stackoverflow.com/questions/36753573/how-do-i-exclude-files-from-git-ls-files

**Reply by zaripych** [reply]
And the fork https://github.com/zaripych/turborepo/commit/4b3d7bbb11f628c7ae68cba39dcef7932a025538 - same changes on top of latest stable version of `turborepo` seemed to work for me locally, but obviously needs more unit tests ...

**Reply by weyert** [reply]
Fascinating never remotely thought of/would have tried :!: :)
Thanks for the fork. I still haven’t cracked the problem of getting turbo binaries of forks to CI

**Comment by nathanhammond** [comment]
Officially, no, we don't support it. However, if you trace through certain places in the codebase, you will see that our glob handling is inconsistent and you can probably sneak them in￼.￼ I don't recommend it as the availability is incidental, but because we are as frustrated as we are with our existing setup it is on our list to rework.￼ Feature request received. 😊

**Reply by jkjustjoshing** [reply]
@nathanhammond what's the status of this? Why was this closed?

**Reply by jkjustjoshing** [reply]
According to https://github.com/vercel/turbo/pull/4066, this has been implemented? However there's no documentation around it. https://turbo.build/repo/docs/reference/configuration#inputs should have a mention of the support, or at least an example in one of the `inputs` arrays.
